ztools/telnet/telnet.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * ZGrab Copyright 2015 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package telnet
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net"
+)
+
+// RFC 854 - https://tools.ietf.org/html/rfc854
+const (
+	IAC                = byte(255) //Interpret as command
+	DONT               = byte(254)
+	DO                 = byte(253)
+	WONT               = byte(252)
+	WILL               = byte(251)
+	GO_AHEAD           = byte(249) // Special go ahead command
+	IAC_CMD_LENGTH     = 3         // IAC commands take 3 bytes (inclusive)
+	READ_BUFFER_LENGTH = 8192
+)
+
+func GetTelnetBanner(logStruct *TelnetLog, conn net.Conn, maxReadSize int) (err error) {
+	if err = NegotiateOptions(logStruct, conn); err != nil {
+		return err
+	}
+
+	//grab banner
+	buffer := make([]byte, READ_BUFFER_LENGTH)
+
+	numBytes := len(buffer)
+	rounds := int(math.Ceil(float64(maxReadSize) / READ_BUFFER_LENGTH))
+	count := 0
+	for numBytes != 0 && count < rounds && numBytes == READ_BUFFER_LENGTH {
```

Comment:
I think we have flipped understandings of maxReadSize and READ_BUFFER_LENGTH. maxReadSize is the total number of bytes that are configured to be read, and READ_BUFFER_LENGTH is the length of the buffer passed to `conn.Read`. The number of rounds is the way that attempt to read READ_BUFFER_LENGTH bytes no more than `max_read_size / READ_BUFFER_LENGTH  + 1` times (except in the case of the last round, where we only want to read the remainder `max_read_size % READ_BUFFER_LENGTH` number of bytes) . The check for numBytes != 0 is to not process the banner if there is no data that has been read (which can happen even without a io.EOF err), and we also consider `conn.Read` to be finished if it doesn't fill up the full READ_BUFFER_LENGTH (i.e. only a partial buffer's worth of data has been returned by the server)